swim/heal_partition.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import "time"
+
+// AttemptHeal attempts to heal a partition between the node and the target.
+//
+// Be mindfull that calling this function will not result in a heal when there
+// are nodes that need to be reincarated to take precedence over the faulty
+// declarations that occur during a network partition. A cluster may therefore
+// need multiple calls to this function with some time in between to heal.
+//
+// Check out ringpop-common/docs for a full description of the algorithm.
+func AttemptHeal(node *Node, target string) error {
+	node.emit(AttemptHealEvent{})
+	node.logger.WithField("target", target).Info("attempt heal")
+
+	// If join request succeeds a partition is detected,
+	// this node will now coordinate the healing mechanism.
+	joinRes, err := sendJoinRequest(node, target, time.Second)
+	if err != nil {
+		return err
+	}
+
+	A := node.disseminator.FullSync()
+	B := joinRes.Membership
+	var csForA, csForB []Change
+
+	// Find changes that are alive for B and faulty for A and visa versa.
```

Comment:
I think it would be easier if you can refactor this loop out to a separate function, which you could unit-test separately. Since this is side-effect free behavior, it will be easier to read the actual logic of partition healing.